### ISSUE TITLE
kubectl get pods.v1beta1.metrics.k8s.io fails, but should succeed

### ISSUE THREAD

**Issue by smarterclayton** [kubernetes/kubernetes#65718]
Metrics v1beta1 resources can't be listed by kubectl 1.10.  It's probably an artifact of kind vs resource split, but this is broken and should work (this is a valid aggregated api with a valid discovery doc).

```
$ kubectl get pods.v1beta1.metrics.k8s.io
error: the server doesn't have a resource type "pods"
$ kubectl describe pods.v1beta1.metrics.k8s.io
error: the server doesn't have a resource type "pods"
$ kubectl get --raw /apis/metrics.k8s.io/v1beta1/
{"kind":"APIResourceList","apiVersion":"v1","groupVersion":"metrics.k8s.io/v1beta1","resources":[{"name":"nodes","singularName":"","namespaced":false,"kind":"NodeMetrics","verbs":["get","list"]},{"name":"pods","singularName":"","namespaced":true,"kind":"PodMetrics","verbs":["get","list"]}]}
```

Expected:

kubectl get should have listed the types by trying to get custom columns for the metrics endpoint (which will be the next bug) or fall back to default display output.  This violates our rules that well formed aggregated APIs should be automatically be listable or describable via the CLI.

I consider this a 1.12 release blocker since this is a beta api that is part of the core

@kubernetes/sig-cli-bugs @kubernetes/sig-autoscaling-bugs

**Comment by smarterclayton** [comment]
@juanvallejo for us to consider metrics official we need to be able to do this via the CLI.  Can you help @DirectXMan12 track this down?

**Comment by liggitt** [comment]
what does `kubectl api-resources --api-group=metrics.k8s.io` show?

**Comment by DirectXMan12** [comment]
```shell
$ k api-resources --api-group=metrics.k8s.io
NAME      SHORTNAMES   APIGROUP         NAMESPACED   KIND
nodes                  metrics.k8s.io   false        NodeMetrics
pods                   metrics.k8s.io   true         PodMetrics
```

**Comment by DirectXMan12** [comment]
that said, I think the issue *might* be a restmapper ignoring discovery information.  I just lost my place debugging, but more info in a couple of minutes.

**Comment by DirectXMan12** [comment]
Yeah, if discovery says the resource name is `pods`, the RESTMapper shouldn't have

```go
(dlv) p m.defaultGroupVersions[0].Group
"metrics.k8s.io"
(dlv) p m.resourceToKind
map[k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/schema.GroupVersionResource]k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/runtime/schema.GroupVersionKind [
        {Group: "metrics.k8s.io", Version: "v1beta1", Resource: "nodemetricslist"}: {
                Group: "metrics.k8s.io",
                Version: "v1beta1",
                Kind: "NodeMetricsList",},
        {Group: "metrics.k8s.io", Version: "v1beta1", Resource: "podmetricslist"}: {
                Group: "metrics.k8s.io",
                Version: "v1beta1",
                Kind: "PodMetricsList",},
        {Group: "metrics.k8s.io", Version: "v1beta1", Resource: "podmetricslists"}: {
                Group: "metrics.k8s.io",
                Version: "v1beta1",
                Kind: "PodMetricsList",},
        {Group: "metrics.k8s.io", Version: "v1beta1", Resource: "nodemetrics"}: {
                Group: "metrics.k8s.io",
                Version: "v1beta1",
                Kind: "NodeMetrics",},
        {Group: "metrics.k8s.io", Version: "v1beta1", Resource: "nodemetricses"}: {
                Group: "metrics.k8s.io",
                Version: "v1beta1",
                Kind: "NodeMetrics",},
        {Group: "metrics.k8s.io", Version: "v1beta1", Resource: "nodemetricslists"}: {
                Group: "metrics.k8s.io",
                Version: "v1beta1",
                Kind: "NodeMetricsList",},
        {Group: "metrics.k8s.io", Version: "v1beta1", Resource: "podmetrics"}: {
                Group: "metrics.k8s.io",
                Version: "v1beta1",
                Kind: "PodMetrics",},
        {Group: "metrics.k8s.io", Version: "v1beta1", Resource: "podmetricses"}: {
                Group: "metrics.k8s.io",
                Version: "v1beta1",
                Kind: "PodMetrics",},
        {Group: "metrics.k8s.io", Version: "v1beta1", Resource: "list"}: {
                Group: "metrics.k8s.io",
                Version: "v1beta1",
                Kind: "List",},
        {Group: "metrics.k8s.io", Version: "v1beta1", Resource: "lists"}: {
                Group: "metrics.k8s.io",
                Version: "v1beta1",
                Kind: "List",},
```


Also, ![filthy hobitses](https://media.giphy.com/media/QZ59vM1qoPRSM/giphy.gif)

**Comment by liggitt** [comment]
does kubectl 1.11 work properly? I know it started leaning more heavily on discovery

**Comment by DirectXMan12** [comment]
That was my thought.  This is testing with a master kubectl

**Comment by DirectXMan12** [comment]
Let me see if I can track down what's going wrong.

**Comment by DirectXMan12** [comment]
ahhh

**Comment by DirectXMan12** [comment]
This guess is wrong:

https://github.com/kubernetes/kubernetes/blob/dd97a7bc59c913c71730dba8b8de274820466417/staging/src/k8s.io/client-go/restmapper/discovery.go#L104

Specifically, it's probably wrong to *completely* ignore the provided plural name if we don't have a singular name.  Patching that (temporarily) to just not have a singular name seems to fix the issue (partially, the 1.12 client seems to be sending an `Accept` header that's being rejected by default, unless you `-o json`).

The solution is probably to just have metrics-server set the proper `singularName` field.  Don't know why the old accept header parsing logic is rejecting the new accept header, though.

**Comment by DirectXMan12** [comment]
RE: the `Accept` header, the header is `application/json;as=Table;v=v1beta1;g=meta.k8s.io, application/json`.  metrics-server is current using code from Kubernetes 1.8 for the apiserver library (currently updating it, along with a big refactor) :-/

**Comment by juanvallejo** [comment]
@smarterclayton 
> for us to consider metrics official we need to be able to do this via the CLI. Can you help 

Sure, will look into this

**Comment by smarterclayton** [comment]
Yikes!

**Comment by juanvallejo** [comment]
@DirectXMan12 
> (partially, the 1.12 client seems to be sending an Accept header that's being rejected by default, unless you -o json)

Hm, if the server is old enough it might be the case that it does not accept as=Table

**Comment by juanvallejo** [comment]
cc @soltysh

**Comment by krishicks** [comment]
@liggitt 1.11 does not work without `-o json`. This is talking to a 1.8.10 server.

**Comment by liggitt** [comment]
> @liggitt 1.11 does not work without -o json. This is talking to a 1.8.10 server

 Correct. Server versions prior to 1.9 had a bug that would not let them respond correctly to requests with Accept headers including mime types they did not support. This was fixed in 1.9

**Comment by deads2k** [comment]
fixed by https://github.com/kubernetes/kubernetes/pull/66249
